cpp/include/cudf/strings/string_view.cuh

```diff
@@ -0,0 +1,323 @@
+ * within a strings column. This is likely created dynamically and temporarily.
+ * It is not recommended to be allocated directly on the global memory heap.
+ *---------------------------------------------------------------------------**/
+class string_view
+{
+ public:
+  string_view() = default;
+  /**---------------------------------------------------------------------------*
+   * @brief Create instance from existing device char array.
+   *
+   * @param data Device char array encoded in UTF8.
+   * @param bytes Number of bytes in data array.
+   *---------------------------------------------------------------------------**/
+  __host__ __device__ string_view(const char* data, size_type bytes);
+  /**---------------------------------------------------------------------------*
+   * @brief Create instance from existing device char array. The array must
+   * include a null-terminator ('\0).
+   *
+   * @param data Device char array encoded in UTF8.
+   *---------------------------------------------------------------------------**/
+  __device__ string_view(const char* data);
+  string_view(const string_view&) = default;
+  string_view(string_view&&) = default;
+  ~string_view() = default;
+  string_view& operator=(const string_view&) = default;
+  string_view& operator=(string_view&&) = default;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Return the number of bytes in this string
+   *---------------------------------------------------------------------------**/
+  __host__ __device__ size_type size() const;
+  /**---------------------------------------------------------------------------*
+   * @brief Return the number of bytes in this string
+   *---------------------------------------------------------------------------**/
+  __host__ __device__ size_type length() const;
+  /**---------------------------------------------------------------------------*
+   * @brief Return the number of characters (UTF-8) in this string
+   *---------------------------------------------------------------------------**/
+  __device__ size_type characters() const;
+  /**---------------------------------------------------------------------------*
+   * @brief Return a pointer to the internal device array
+   *---------------------------------------------------------------------------**/
+  __host__ __device__ const char* data() const;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Return true if string has no characters
+   *---------------------------------------------------------------------------**/
+  __device__ bool empty() const;
+  __device__ bool is_null() const;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Handy iterator for navigating through encoded characters.
+   *---------------------------------------------------------------------------**/
+  class iterator
```

Comment:
If we're going to have this, it should be a standard's compliant iterator with the appropriate type definitions and traits. It should also be a `const_iterator` since it doesn't allow modifying the data.

Reply:
I've not found any good documentation on building a standard compliant iterator. I could use some help on this.

Reply:
Perhaps this should be moved out to a separate issue and PR?

Reply:
I was able to figure out how to get the iterator to work with `thrust::copy` at least.